Extend prevent-late-fallback by lock-counter

zentax-dev opened the pull request:
This feature allows to lock the the ``remaining_attempts`` counter.
When ``remaining_attempts`` is locked, the bootloader should not decremented and  incremented the variable anymore during each boot.
It is active when a slot is _marked good_ and inactive when a slot is _marked active_.
This way it prevents fallback to an earlier version, whilst inhibiting additional write cycles to the target medium. 
The status can be printed out with `barebox-state` and `rauc status`.
In a prelimary talk with @ejoerns the decision was made to not add this to the _D-Bus_ interface yet. It will be added in a future pull request.
This feature also needs to be supported by the bootloader. 
So far, a patch has been handed in for _barebox_ to support this feature, see [](https://lists.infradead.org/pipermail/barebox/2025-June/051393.html)

Commit:
config_file: extend prevent-late-fallback with lock-counter

In general the new option for prevent late fallback adds the possibility
to lock the attempts counter.

As first step towards this, the parser is extended to support the new value.

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

Commit:
src/main: add counter locking info to rauc status output

This adds the option to print the current state of prevent-late-fallback.
The possibility to get/set it via D-Bus is not yet
implemented and will be added separately.

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

Commit:
src/context: check for lock-counter inconsistency

It is best to show problems with inconsistent configuration early.
When counter locking is enabled in rauc, it must also be set in the
bootloader.

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

Commit:
docs/reference: add info about lock-counter

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

ejoerns commented on src/main.c:
```diff
@@ -2166,6 +2174,19 @@ static gboolean status_start(int argc, char **argv)
 		}
 
+		/* This checks for inconsistency between configured option in rauc config
+		 * and the current state of the required barebox variable. So  */
```
Should the sentence continue here?

ejoerns commented on src/main.c:
```diff
@@ -2166,6 +2174,19 @@ static gboolean status_start(int argc, char **argv)
+		gboolean locked = FALSE;
+		if (r_context()->config->prevent_late_fallback == R_CONFIG_FALLBACK_LOCK_COUNTER) {
+			if (!r_boot_get_lock_counter(&locked, &ierror)) {
+				g_printerr("Failed to determine prevent-late-fallback state: %s\n", ierror->message);
```
```suggestion
				g_printerr("Failed to determine boot counter lock state: %s\n", ierror->message);
```

ejoerns commented on src/main.c:
```diff
@@ -2166,6 +2174,19 @@ static gboolean status_start(int argc, char **argv)
+				g_printerr("Failed to determine prevent-late-fallback state: %s\n", ierror->message);
+				g_clear_error(&ierror);
+			} else if (!locked) {
+				g_printerr("Inconsistency lock-counter state: enabled in rauc but not in bootloader");
```
```suggestion
				g_printerr("Inconsistency in 'lock-counter' handling: enabled, but not supported by the bootloader");
```

ejoerns commented on src/mark.c:
```diff
@@ -153,8 +155,15 @@ gboolean r_mark_active(RaucSlot *slot, GError **error)
 	slot_state->activated_timestamp = g_date_time_format(now, "%Y-%m-%dT%H:%M:%SZ");
-	slot_state->activated_count++;
 
+	if (r_context()->config->prevent_late_fallback == R_CONFIG_FALLBACK_LOCK_COUNTER) {
```
Is there a specific reason for placing it between two `slot_state` variable updates?

Another potential issue might be that if `r_boot_set_primar()` succeeds but `r_boot_set_lock_counter()` fails for some reason, you might end in a system that is locked to the new slot, i.e. which does not provide a fallback anymore. So maybe changing the order of the  `r_boot_set_lock_counter()` and `r_boot_set_primar()` call might make sense here?

This would also be consistent with the order used in `r_mark_good()` below because the locking then surrounds the other bootloader operations.

ejoerns replied:
I guess this was changed and discussed in the latest version. Will mark it as 'resolved', though.

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+    'rauc status mark-bad other' after marking the currently booted slot as good.
+    This means that the other slot(s) is/are no longer eligible for fallback.
+    It behaves exactly like the *deprecated* ``prevent-late-fallback=false``.
+  * ``lock-counter`` This is a slot independent setting that preserves
```
`prevent-late-fallback` is always slot-independent, no matter which value you give.

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+            };
+    };
+
+  With this configuration, the bootloader is aware of the locking state due to the
```
```suggestion
  With this configuration, barebox is aware of the locking state due to the
```

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+  With this configuration, the bootloader is aware of the locking state due to the
+  usage of the variable in the devicetree. So the bootloader stops decrementing
+  the ``remaining_attempts`` counter.
+  Slots locking is activated when 'rauc status mark-good' is executed
```
```suggestion
  Boot counter locking is activated when 'rauc status mark-good' is executed
```

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+  Slots locking is activated when 'rauc status mark-good' is executed
+  and deactivated when 'rauc status mark-active' is executed. This can happen either
+  implicitly during a RAUC update or explicitly through manual commands.
+  The current value of the ``slots_locked`` variable can be read via the commands
```
```suggestion
  The current value of the ``attempts_locked`` state variable can be read via the commands
```

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+  and deactivated when 'rauc status mark-active' is executed. This can happen either
+  implicitly during a RAUC update or explicitly through manual commands.
+  The current value of the ``slots_locked`` variable can be read via the commands
+  ``barebox-state -d`` or and ``rauc status``.
```
```suggestion
  ``barebox-state -d`` and ``rauc status`` (without service).
```

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+  implicitly during a RAUC update or explicitly through manual commands.
+  The current value of the ``slots_locked`` variable can be read via the commands
+  ``barebox-state -d`` or and ``rauc status``.
+  It is not yet available via die D-Bus interface.
```
```suggestion
  It is not yet available via the D-Bus interface and thus cannot be displayed when calling `rauc status` with the service enabled.
```

zentax-dev replied:
nope wasn't intentional

ejoerns commented on src/config_file.c:
```diff
@@ -965,8 +987,17 @@ gboolean load_config(const gchar *filename, RaucConfig **config, GError **error)
 
 	c->prevent_late_fallback = g_key_file_get_boolean(key_file, "system", "prevent-late-fallback", &ierror);
```
This now assigns a boolean to an enum; if this is intentional, it might at least deserve a comment in code.

And, if we deprecate the boolean values, we should at least print some kind of deprecation info/warning.
I guess this can easily be added in an `else` statement below.

ejoerns replied:
@zentax-dev Thanks for adding the comment.

However, the suggested depreciation warning has not been implemented yet, correct?

zentax-dev replied:
Done. used a `g_info` for now. Tests fail when we use a `g_warning`.
And I think, an info is enough as a starter. Once a decision is made when the deprecated old parameters will be deleted, this should be changed into a warning.

ejoerns reviewed (changes requested):
@zentax-dev Thank you for the contribution!

I had a look at the code and left some comments where I think the feature still deserves some rework. I guess most are just about proper wording/naming and error handling.

You should also point out in the documentation when and why boot counter locking might be preferable to the existing 'mark bad other' solution we have.

ejoerns commented on docs/reference.rst:
```diff
@@ -252,13 +252,53 @@ Example configuration:
+    The syntax has changed. Please use the new values.
+    In order to get the same behaviour set ``prevent-late-fallback=mark-bad``.
+
+``prevent-late-fallback=<disabled/mark-bad/lock-counter>`` (optional)
```
But in our case, no option is deprecated. The values are.

zentax-dev commented on include/bootchooser.h:
```diff
@@ -55,6 +55,28 @@ G_GNUC_WARN_UNUSED_RESULT;
+ *
+ * @return TRUE if successful, FALSE if failed
+ */
+gboolean r_boot_set_lock_counter(gboolean locked, GError **error);
```
Okay, phrased differently then, what would change if we rename this method?
If we swap "counter" and "lock" you do not know, to which variable it belongs to, and you'd have to look that up and probably what it exactly is for. 
In the current state, you can map it to the variable it belongs to easier, but you'd have to look up what it exactly does

ejoerns replied:
> If we swap "counter" and "lock" you do not know, to which variable it belongs to, and you'd have to look that up and probably what it exactly is for.

I can't see how this belongs to any variable. Could you point out which variable you refer to?

ejoerns replied:
FTR: We have clarified this in chat and it's implemented as proposed now (as `r_boot_set_counters_lock()`)

a3f commented on src/mark.c:
```diff
@@ -141,6 +143,13 @@ gboolean r_mark_active(RaucSlot *slot, GError **error)
 	r_slot_status_load(slot);
 	slot_state = slot->status;
 
+	if (r_context()->config->prevent_late_fallback == R_CONFIG_FALLBACK_LOCK_COUNTER) {
```
Please add a comment explaining why locking removal happens before set_primary here.

We can set multiple state variables at once atomically, but we don't. This might very well not be a problem, but it's not immediately clear why. A comment should help explaining why a powercut between removal of lock and set_primar below doesn't harm correctness.

a3f commented on src/mark.c:
```diff
@@ -177,6 +185,13 @@ gboolean r_mark_good(RaucSlot *slot, GError **error)
 		return FALSE;
 	}
 
+	if (r_context()->config->prevent_late_fallback == R_CONFIG_FALLBACK_LOCK_COUNTER) {
```
Likewise, a comment is needed here, why this order was chosen.

a3f commented on src/mark.c:
```diff
@@ -170,13 +182,23 @@ gboolean r_mark_good(RaucSlot *slot, GError **error)
 	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
 
+	/* It is more sensible to mark the slot as good before locking the counter. As in case of a power fail between
+	 * marking and locking, we'd have a system with a working fallback to the old stot, so in case of problems
```
```suggestion
	 * marking and locking, we'd have a system with a working fallback to the old slot, so in case of problems
```

I suggest pointing out that on next boot, the mark good still has a chance to lock, so it's eventually consistent of sorts. Also I think comment style used differs from the usual RAUC convention. Please align.

a3f reviewed (commented):
I have no further comments on the series and haven't looked deep enough to put a checkmark on it.

zentax-dev commented:
> I have no further comments on the series and haven't looked deep enough to put a checkmark on it.

I'll mark your remarks as resolved then

zentax-dev commented:
@ejoerns would be nice if you could review this again

js731ca commented:
gave this PR + the corresponding barebox changes a spin - nice work!
sth i stumbled into: rauc fails when configured with `prevent-late-fallback=lock-counter`, but without the barebox-state being available for $reasons
-> could this be handled more gracefully by maybe falling back to 'disabled'?

js731ca commented:
is this a minor bug/regression? - the 'rauc status' always reports `Counter lock: disabled`, or what is the meaning of that status?
<details>

```
$> barebox-state -d
bootstate.systemA.remaining_attempts=3
bootstate.systemA.priority=10
bootstate.systemB.remaining_attempts=3
bootstate.systemB.priority=20
bootstate.last_chosen=2
bootstate.attempts_locked=1
```
```
$> rauc status
=== System Info ===
Compatible:  Test System
Variant:     abc,xyz-rev-c
Booted from: fit.1 (systemB)
Counter lock: disabled

=== Bootloader ===
Activated: fit.1 (systemB)
```
</details>

js731ca commented on src/main.c:
```diff
@@ -1540,7 +1541,8 @@ static gchar* r_status_formatter_readable(RaucStatusPrint *status)
 	g_string_append_printf(text, "Variant:     %s\n", status->variant);
-	g_string_append_printf(text, "Booted from: %s (%s)\n\n", bootedfrom ? bootedfrom->name : NULL, status->bootslot);
+	g_string_append_printf(text, "Booted from: %s (%s)\n", bootedfrom ? bootedfrom->name : NULL, status->bootslot);
+	g_string_append_printf(text, "Counter lock: %s\n\n", status->lock_counter ? "enabled" : "disabled");
```
```suggestion
//	g_string_append_printf(text, "Counter lock: %s\n\n", status->lock_counter ? "enabled" : "disabled");
```
a `rauc status` depends on `retrieve_status_via_dbus` - which isn't implemented yet
so the status would always say "disabled" - which is misleading
(and mis-lead me to debug into the code... sth we could save others from ;-)

ejoerns replied:
And of course, this should only be printed when counter locking is enabled!

zentax-dev replied:
`r_status_formatter_readable` is called via `print_status` -> `status_start`
And there `print_status` is just called if `!ENABLE_SERVICE`
Or is there another way to call it?

js731ca commented:
> But I think you mean the case where you have A/B boot slots, but only a single root slot?

when everything is A/B (bootloader, fitimage/kernel, rootfs); during an update rauc only installs a new bootloader and does an atomic operation (to switch the emmc boot-part select bit). Note that neither rauc nor barebox flip that switch back when one slot depletes attempts.
The "problem"/cornercase we would run into if the boot-slot is switched back externally.

js731ca commented:
> @ejoerns would be nice if you could review this again

and @a3f : gentle(?) ping :-)
the barebox parts where merged with 2025.06.01 - but since both parts are involved: the more reviewers the better

zentax-dev commented:
> > @ejoerns would be nice if you could review this again
> 
> and @a3f : gentle(?) ping :-) the barebox parts where merged into master - but since both parts are involved: the more reviewers the better

Just a quick heads up: I've briefly talked to @ejoerns about this and he will look into it.
As far as I know @a3f was already alright with everything, but I'll see him the next days, so I'll have a chat with him :)

ejoerns commented:
> > But I think you mean the case where you have A/B boot slots, but only a single root slot?
> 
> when everything is A/B (bootloader, fitimage/kernel, rootfs); during an update rauc only installs a new bootloader and does an atomic operation (to switch the emmc boot-part select bit). Note that neither rauc nor barebox flip that switch back when one slot depletes attempts. The "problem"/cornercase we would run into if the boot-slot is switched back externally - and happens to be out-of-sync with the proper grouping

Is that an actual use case or a theoretical consideration?

For most cases, the update of the bootloader is considered atomic, but without any fallback.

A proper fallback would require an entity before the bootloader that can decide that a fallback is necessary.

A design that actually expects the bootloader to be part of the A/B setup would also need to ensure that the bootloader boots the corresponding slot. Thus, if there is actually a fallback, the system with the older RAUC version should be booted instead. This sounds like what #1750 targets. For good reasons, the boot partition switching is not performed by RAUC here (but by the bootloader itself instead).

ejoerns commented:
> > gave this PR + the [corresponding barebox changes](https://lore.kernel.org/barebox/20250619-bootchooser-lock-v3-0-fc4ad92c3866@pengutronix.de/) a spin - nice work! sth i stumbled into: rauc fails when configured with `prevent-late-fallback=lock-counter`, but without the barebox-state being available for $reasons
> > ```
> > rauc[839]: Using central status file /mnt/data/rauc/central.raucs
> > rauc[839]: Using system config file /etc/rauc/system.conf
> > rauc[845]: no such variable: bootstate.attempts_locked
> > rauc[839]: Failed to initialize context: No content to read
> > systemd[1]: rauc.service: Failed with result 'exit-code'.
> > systemd[1]: Failed to start RAUC Update Service.
> > ```
> >   
> > -> could this be handled more gracefully by maybe falling back to 'disabled'?
> 
> The intention behind this is to make it clear, that the system is misconfigured and not working as intended. If rauc would fall back to disabled, the service would be running, and the user probably wouldn't notice it

At least the error checking and reporting could be improved here, since the message actually does not give a good indication about what's going wrong. The part that could be addressed in this PR is a change in `src/context.c` to add something like

``` diff
                        if (!r_barebox_get_lock_counter(&locked, &ierror)) {
-                               g_propagate_error(error, ierror);
+                               g_propagate_prefixed_error(error, ierror, "Failed to read barebox lock counter: ");
                                return FALSE;
```

The other part is that output parsing and error code checking for barebox-state happen in an unoptimal order currently. But that's the case for existing methods, too.

ejoerns commented:
> The other part is that output parsing and error code checking for barebox-state happen in an unoptimal order currently. But that's the case for existing methods, too.

For this, I've created https://github.com/rauc/rauc/pull/1781 now.

ejoerns commented on include/bootchooser.h:
```diff
@@ -55,6 +55,28 @@ G_GNUC_WARN_UNUSED_RESULT;
+ * Check current status of boot counter locking in the bootloader.
+ * This variable is not set per slot, but is valid for all slots.
+ *
+ * @param locked return location for value of lock-counter
```
What's `lock-counter` here? So far this is only the name of the 'prevent-late-fallback' strategy.

zentax-dev replied:
hmhm, to keep it generic and not just to match barebox: it is a related bootloader value
so maybe something like
```suggestions
 * @param locked return location for the current value in the bootloader
```

ejoerns commented on include/bootchooser.h:
```diff
@@ -55,6 +55,28 @@ G_GNUC_WARN_UNUSED_RESULT;
+
+/**
+ * Check current status of boot counter locking in the bootloader.
+ * This variable is not set per slot, but is valid for all slots.
```
Maybe just drop this line.

ejoerns commented on src/bootchooser.c:
```diff
@@ -155,6 +156,61 @@ gboolean r_boot_set_state(RaucSlot *slot, gboolean good, GError **error)
+	GError *ierror = NULL;
+	gboolean res = FALSE;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
```
```suggestion
    g_return_val_if_fail(locked != NULL, FALSE);
	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
```

ejoerns commented on src/mark.c:
```diff
@@ -1,6 +1,8 @@
+#include "config_file.h"
 #include "event_log.h"
 #include "context.h"
+#include "glib.h"
```
See above.

ejoerns commented on src/mark.c:
```diff
@@ -141,6 +143,17 @@ gboolean r_mark_active(RaucSlot *slot, GError **error)
 	r_slot_status_load(slot);
 	slot_state = slot->status;
 
+	/* It is more sensible to unlock the counter, before marking it primary. Reversing the order would create
```
```suggestion
	/* It is more sensible to unlock the counter before marking it primary. Reversing the order would create
```

For users not aware of the previous implementation attempt, the 'It is more sensible' sounds a bit unexpected. We also don't mark the counter primary.
So maybe just start with `Unlock boot counting before marking the slot primary...`.

ejoerns commented on src/mark.c:
```diff
@@ -141,6 +143,17 @@ gboolean r_mark_active(RaucSlot *slot, GError **error)
+	/* It is more sensible to unlock the counter, before marking it primary. Reversing the order would create
+	 * an unintended corner case where a power cut could leave the system configured to boot from a new,
+	 * untested slot without the ability to fall back if that slot fails to boot properly, as it would
+	 * still be locked. So we prepary the safety mechanisms first and then do the potentially risky operation. */
```
```suggestion
	 * still be locked. So we prepare the safety mechanisms first and then do the potentially risky operation. */
```

'safe' and 'risky' sounds a bit inappropriate for the individual operations.

zentax-dev replied:
you're right, and also the information probably is also a bit redundant when taking the previous sentence into account.
```suggestion
	 * still be locked. So it is better to prepare the fallback mechanisms first. */
```

ejoerns commented on src/mark.c:
```diff
@@ -170,13 +182,24 @@ gboolean r_mark_good(RaucSlot *slot, GError **error)
 	g_return_val_if_fail(slot, FALSE);
 	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
 
+	/* It is more sensible to mark the slot as good before locking the counter. As in case of a power fail between
```
Above, this section is attached to the `R_CONFIG_FALLBACK_LOCK_COUNTER` check, which also seems more appropriate here since for 90% of the cases, the `r_boot_set_state()` will not have anything to do with slot locking.

ejoerns commented on src/mark.c:
```diff
@@ -230,7 +253,7 @@ gboolean mark_run(const gchar *state,
 		}
 
-		if (r_context()->config->prevent_late_fallback
+		if (r_context()->config->prevent_late_fallback == R_CONFIG_FALLBACK_MARK_BAD
```
This should be part of the commit changing the type of `prevent_late_fallback`.

ejoerns commented on src/main.c:
```diff
@@ -1540,7 +1541,8 @@ static gchar* r_status_formatter_readable(RaucStatusPrint *status)
 	g_string_append_printf(text, "Compatible:  %s\n", status->compatible);
 	g_string_append_printf(text, "Variant:     %s\n", status->variant);
-	g_string_append_printf(text, "Booted from: %s (%s)\n\n", bootedfrom ? bootedfrom->name : NULL, status->bootslot);
+	g_string_append_printf(text, "Booted from: %s (%s)\n", bootedfrom ? bootedfrom->name : NULL, status->bootslot);
```
In the commit message:

```
    src/main: add prevent-late-fallback to userspace tools
```

Since RAUC only runs in 'userspace', I'd leave that term out. The commit also doesn't add 'prevent-late-fallback', it just adds output for counter locking. Essentially, you `add counter locking information to rauc status output`.

```
    This adds the option to print the current state of prevent-late-fallback
    and also to enable and disable sets the lock-counter via userspace.
```

The sentence looks inconsistent. Also, the corresponding commit does not add any functionallity to *enable/disable/set* something.

```
    The possibility to get/set it via D-Bus is not yet
    implemented and will be added separately.
```

ejoerns commented on test/bootchooser.c:
```diff
@@ -274,7 +284,7 @@ static void bootchooser_barebox_conf_attempts(BootchooserFixture *fixture,
 	GError *error = NULL;
 	gboolean res;
-
+	// TODO[lsc]: add test here
```
This is probably still open?

zentax-dev commented on src/bootloaders/barebox.c:
```diff
@@ -199,6 +206,105 @@ gboolean r_barebox_set_state(RaucSlot *slot, gboolean good, GError **error)
 	return TRUE;
 }
 
+/* This freezes the remaining attempts counter in barebox by setting the attempts_locked variable */
```
or maybe better "```suggestion
/* This freezes the remaining attempts counter of the currently activated slot in barebox by setting the attempts_locked variable */
```

Commit:
barebox: add counter locking to barebox backend

Forward the counter locking option to the barebox bootloader.
So the barebox can then stop decrementing the remaining_attempts
counter.
It inhibits fall-back to a previous version of the system, which
can happen if a system is rebootet too frequently before a slot
is marked good again and the remaining_attempts counter is
incrememented. As a side effect, it inhibits excessive write
cycles on the storage medium.

This also needs changes to barebox, see [1].

[1] https://lists.infradead.org/pipermail/barebox/2025-June/051393.html

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

Commit:
bootchooser: add counter-locking to bootchooser

The setting is currently only supported by barebox.

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

a3f commented:
We can't necessarily expect that an older bundle is never installed.
FOTA is not the only one way to do an update, USB recovery (with a bundle that often lags behind the online version) is a possibility in my project. The problematic sequence could look as follows:

- FOTA installation of new RAUC bundle with this feature here
- System is partially broken: boot-complete is reached, but it's not possible to do further FOTA
- USB recovery stick is inserted, update to old system: booth rootfs and bootloader are at old state now
- Recovered system fails to start up
- As old system doesn't have support for switching eMMC boot partition on attempt depletion, we end up booting the new rootfs with the old bootloader
- RAUC refuses to startup: online USB recovery no longer possible

With some more thought put into this contrived example, one might be able to arrive at a simpler sequence of events.

With a recovery mode implemented directly into the bootloader we side step this particular issue, but it doesn't sit well with me that there are sequences that have the potential for RAUC to fail to startup not because of prior misconfiguration, but because of the bootloader version in the field. I'd rather have RAUC warn very loudly about this, but then do some fallback behavior, either set the attempts to a high number or just ignore the locking.

Interested in your thoughts.

Commit:
mark: make use of counter locking

When activated, the attempts counter of the slot that is marked good will
be locked and not decrease and increase anymore.
It will be unlocked again, when the slot is marked active.

Signed-off-by: Lars Schmidt <l.schmidt@pengutronix.de>

zentax-dev commented on src/context.c:
```diff
@@ -404,6 +405,30 @@ static gboolean r_context_configure_target(GError **error)
+			gboolean locked = FALSE;
+			if (!r_barebox_get_lock_counter(&locked, &ierror)) {
+				g_propagate_prefixed_error(error, ierror, "Failed to read barebox lock counter: ");
+				return FALSE;
```
Thanks for the hint, I've changed this and added a test for it
